Rules/UseCorrectCasing.cs

```diff
@@ -0,0 +1,176 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Management.Automation.Language;
+using Microsoft.Windows.PowerShell.ScriptAnalyzer.Generic;
+#if !CORECLR
+using System.ComponentModel.Composition;
+#endif
+using System.Globalization;
+using System.Linq;
+using System.Management.Automation;
+
+namespace Microsoft.Windows.PowerShell.ScriptAnalyzer.BuiltinRules
+{
+    /// <summary>
+    /// UseCorrectCasing: Check if cmdlet is cased correctly.
+    /// </summary>
+#if !CORECLR
+    [Export(typeof(IScriptRule))]
+#endif
+    public class UseCorrectCasing : ConfigurableRule //IScriptRule
+    {
+        /// <summary>
+        /// AnalyzeScript: Analyze the script to check if cmdlet alias is used.
+        /// </summary>
+        public override IEnumerable<DiagnosticRecord> AnalyzeScript(Ast ast, string fileName)
+        {
+            if (ast == null) throw new ArgumentNullException(Strings.NullAstErrorMessage);
+
+            IEnumerable<Ast> commandAsts = ast.FindAll(testAst => testAst is CommandAst, true);
+
+            // Iterates all CommandAsts and check the command name.
+            foreach (CommandAst commandAst in commandAsts)
+            {
+                string commandName = commandAst.GetCommandName();
+
+                // Handles the exception caused by commands like, {& $PLINK $args 2> $TempErrorFile}.
+                // You can also review the remark section in following document,
+                // MSDN: CommandAst.GetCommandName Method
+                if (commandName == null)
+                {
+                    continue;
+                }
+
+                using (var powershell = System.Management.Automation.PowerShell.Create())
```

Comment:
I would be much happier to use a single instance of PowerShell rather than to create a new instance of PowerShell for each found command.
Also, what do we do about things which look like a command, but can't be found by Get-Command. Should we attempt to validate those?

Reply:
I changed it to use the `Helper.Instance.CommandInfo` method for the moment because this method has commandinfo caching (but the disadvantage of this method is that it has uses a lock internally).
Or maybe would it be better to just create and cache PowerShell in this rule on its own and use the async Execution APIs?

Reply:
I reckon for the moment this is ok - it's something that I'll watch though.